lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemFactory.java

```diff
@@ -120,12 +126,21 @@ public S3FileSystemFactory(OpenTelemetry openTelemetry, S3FileSystemConfig confi
                 config.getSseKmsKeyId(),
                 Optional.empty(),
                 config.getCannedAcl());
+
+        this.uploadExecutor = new ThreadPoolExecutor(
```

Comment:
This doesn't do what you expect. If you very carefully read the Javadoc for `ThreadPoolExecutor`, you'll see that this will result in tasks being rejected when the maximum pool size is reached. An easier way to see why this is true is that there is no queue to hold the non-runnable tasks, so the only possible behavior would be to block the caller or reject the task, and it doesn't block the caller.

If you spend a very long time thinking about `ThreadPoolExecutor` and reading the documentation many times, you may eventually realize that the `cached` and `fixed` are the only useful configurations. There's a reason that we don't use it directly in Trino.

If you want to limit concurrency without having a fixed number of threads, the alternative is to use `BoundedExecutor`, which we do use in various places. Once we can safely use virtual threads, all of this complexity goes away -- just use a `Semaphore`.